### PR TITLE
refactor(authorization): split authorization handlers in different classes

### PR DESCRIPTION
**Pull request by andresmoschini**
Hi @FromDoppler/sap 

It is to try to split the different responsibilities in different authorization handlers, I did not test it manually, and I am not sure if our coverage is enough good, but I am sharing this to allow us to see what I am working on.

### CODE REVIEW

File: Billing.API/DopplerSecurity/IsOwnResourceHandler.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+            return Task.CompletedTask;
+        }
+
+        private bool IsOwnResource(AuthorizationHandlerContext context)
+        {
+            var tokenUserId = context.User.FindFirst(c => c.Type == ClaimTypes.NameIdentifier)?.Value;
+            var resource = context.Resource as Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext;
+
+            if (resource is null)
```
Comment (nreal) [inline Billing.API/DopplerSecurity/IsOwnResourceHandler.cs:32]: Great !!

**Comment by bamaro-dev** [comment]
> > It looks great! If one of those Handlers check the requirement like successful, that it's enough, rigth?
> 
> I suppose that, but I am not sure. Do you know if the integration tests are verifying that scenario?

I think yes, because there are an apikey "isSu" and another that it's not isSu but the ID it's the same as url. Each of those cases should be successful in each of Handlers.

**Comment by rockambole86** [comment]
Maybe silly question... in the case we need to combine one or more (I mean... AND instead of OR) should we put the logic inside a single handler, right?

**Comment by jhoffman-ms** [comment]
excellent, if it works fine maybe we can rename the requirement ```"IsSuperUserOrOwnResourceRequirement" ``` by ``` "AuthorizationRequirement" ```. What you think?

**Comment by andresmoschini** [comment]
> Maybe silly question... in the case we need to combine one or more (I mean... AND instead of OR) should we put the logic inside a single handler, right?

I think that `AND` would be managed by adding requirements. `OR` is managed by adding handlers for the same requirement.

**Comment by andresmoschini** [comment]
> maybe we can rename the requirement "IsSuperUserOrOwnResourceRequirement" by "AuthorizationRequirement". What you think?

In my opinion, each requirement should represent exactly what it _requires_.

For example, we could add a different requirement for download URLs.

**Comment by andresmoschini** [comment]
Hi team, @msimionati has added some improvements in the tests and we have tested it manually.

I have also added a new commit to decouple the handlers from the requirement, so, the binding is done dynamically in the configuration.

Could you take a look and share with me your opinion? I am planning to merge it after lunch.

**Comment by jhoffman-ms** [comment]
For me is Ok. Good job.

**Review by bamaro-dev — COMMENTED** [review]
Looks great!

**Pull request merged**